### PR TITLE
ipq806x: Add support for Netgear Nighthawk Pro Gaming XR500

### PR DESCRIPTION
**Pull request by pgwipeout**
Based on the patch from adamhnat.
https://github.com/openwrt/openwrt/pull/3215/commits/f846f25d93c4f60debefdf870ec8e25c5ba8088e
Cleaned up, rebased, and synced to the existing r7800 support.

Manufacturer page: https://www.netgear.com/gaming/xr500/
Specifications:
Almost identical to Netgear R7800(https://openwrt.org/toh/netgear/r7800)
Differences:
Outside casing
Bigger flash storage
Removed esata
swapped leds: (r7800 > xr500)
usb1 (gpio 7 > 8)
usb2 (gpio 8 > 26)
guest/esata (gpio 26 > 7)
SoC: Qualcomm Atheros IPQ8065
RAM: 512 MB
Storage: 256 MiB
Wireless: Qualcomm Atheros QCA9984
Ethernet: 5x 1000/100/10
USB: 2x 3.0
Install via TFTP recovery:
1.Turn off the power, push and hold the reset button (in a hole on
backside) with a pin
2.Turn on the power and wait till power led starts flashing white (after
it first flashes orange for a while)
3.Release the reset button and tftp the factory img in binary mode. The
power led will stop flashing if you succeeded in transferring the image,
and the router reboots rather quickly with the new firmware.
4.Try to ping the router (ping 192.168.1.1). If does not respond, then
tftp will not work either.
Uploading the firmware image with a TFTP client
$ tftp 192.168.1.1
bin
put openwrt-ipq806x-netgear_xr500-squashfs-factory.img
Signed-off-by: Adam Hnat adamhnat@gmail.com
Signed-off-by: Peter Geis <pgwipeout@gmail.com>

---

I do not have a xr500 to test this, but it is compile tested.
I have tested it on the r7800, with no obvious hardware regressions.

### CODE REVIEW

**Comment by pkgadd** [comment]
Renaming `qcom-ipq8065-r7800.dts` to `qcom-ipq8065-nighthawk.dtsi` (with some changes) seems to be rather pointless, and -as implemented (without adapting the referencing places)- would break building for the r7800.

The last approach of adding xr500 support (imho using a neater way of pushing most of the `*.dts` for r7800 and xr500 into a commonly sourced `*.dtsi`) fizzled out, when trying to get back and forth with the OEM firmware via tftp (hard-)bricked the devices of the (at least two) PR submitters/ testers, so tftp recovery on this device might be a little fragile (warranting a big cover warning). Also consider to structure the commit message a bit differently, following the lead of other devices:
- da8428d277cd3373b05330cb3b4f93aef717c5ab
- c5b44af2fc5ffe620405ccb743b6f4acc461d4de
- d57ba86e489e363be3b26b9cb880048da5dfd12c
- …

**Comment by pgwipeout** [comment]
> Renaming `qcom-ipq8065-r7800.dts` to `qcom-ipq8065-nighthawk.dtsi` (with some changes) seems to be rather pointless, and -as implemented (without adapting the referencing places)- would break building for the r7800.

There is still a qcom-ipq8065-r7800.dts.
The end result is the same dtb file, builds are not affected in any way.
I've tested this on my r7800 (besides, builds would fail otherwise) and build tested the xr500.
The "rename" is just how git chose to interpret the changes, since the resulting common dtsi is ~85% of what either dts was.

Leaving them as separate dts files with no common dtsi is also possible, but as stated there's about 85% commonality.

> 
> The last approach of adding xr500 support (imho using a neater way of pushing most of the `*.dts` for r7800 and xr500 into a commonly sourced `*.dtsi`) fizzled out, when trying to get back and forth with the OEM firmware via tftp (hard-)bricked the devices of the (at least two) PR submitters/ testers, so tftp recovery on this device might be a little fragile (warranting a big cover warning). Also consider to structure the commit message a bit differently, following the lead of other devices:

If you notice, I was one of the reviewers/testers on the last series. I built and had working experimental images from that source.
(That device is still running one of those images)
In regards to the hard brick, adamhnat discovered his was an unrelated power failure:
https://github.com/openwrt/openwrt/pull/3215#issuecomment-702371584
I bounced back and forth several times during that review without issue.

If I had to venture a guess, a quick dump of several oem firmware images shows they seem to be lacking any dts/uboot files. (Though there are files to handle flashing it if it existed)
That leads me to believe they require the existing u-boot to remain intact.
If during testing someone accidentally damaged u-boot, there would be no recovering from that, but that's the case with most of these devices.

The concern about the commit message is valid, I left the original message intact but I can reformat it as necessary.

> 
> * [da8428d](https://github.com/openwrt/openwrt/commit/da8428d277cd3373b05330cb3b4f93aef717c5ab)
> * [c5b44af](https://github.com/openwrt/openwrt/commit/c5b44af2fc5ffe620405ccb743b6f4acc461d4de)
> * [d57ba86](https://github.com/openwrt/openwrt/commit/d57ba86e489e363be3b26b9cb880048da5dfd12c)
> * …
> 
> Also see the following previous PRs:
> 
> * [ipq806x: Add support for Netgear Nighthawk Pro Gaming XR500 #3215](https://github.com/openwrt/openwrt/pull/3215)
> * [IPQ806x: add Netgear xr500 #3052](https://github.com/openwrt/openwrt/pull/3052)
> * [ipq806x: add support for NETGEAR XR500 #2076](https://github.com/openwrt/openwrt/pull/2076)

**Comment by pgwipeout** [comment]
Updated commit message.

File: target/linux/ipq806x/image/Makefile [hunk 1]
```diff
@@ -284,6 +284,22 @@ define Device/netgear_r7800
...
+	DEVICE_VENDOR := NETGEAR
+	DEVICE_MODEL := Nighthawk XR500
+	SOC := qcom-ipq8065
+	KERNEL_SIZE := 4096k
+	NETGEAR_BOARD_ID := XR500
+	NETGEAR_HW_ID := 29764958+0+256+512+4x4+4x4+cascade
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	BOARD_NAME := xr500
+	SUPPORTED_DEVICES += xr500
```
Comment (adschm) [inline target/linux/ipq806x/image/Makefile:298]: Please remove BOARD_NAME and SUPPORTED_DEVICES.
I will have a look at the other stuff later.

**Comment by neheb** [comment]
As this is a new submission, would DSA make sense here to avoid having to convert it later?

**Comment by pkgadd** [comment]
I don't think that is necessary, as the whole switch setup is identical to the r7800 and factored out into a common `*.dtsi` with this PR. 

Yes, merging xr500 support before the dsa migration would incur a merge conflict, but that's relatively easily resolvable as the whole block remains identical to the current r7800 `*.dts`. The same goes for potentially merging the [DSA PR](https://github.com/openwrt/openwrt/pull/4036) first (maybe a bit easier in that order).

**Comment by Ansuel** [comment]
@neheb we really don't know what will be merged first so i would keep this with swconfig and then update the ipq806x pr to convert this new device to dsa

The best case scenario would be 
- merge this 
- merge base qca8k patch
- covert single target to dsa (ipq806x, ath79...)

**Comment by adschm** [comment]
So, now we have essentially reached the same state as with the end of the previous PR.

My facelifted version of this PR can be found here: https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=commitdiff;h=37cffc7e1311b3a66bafb9e5eb9b2d64f700991a

Open question are essentially the same as before:
* Are the MAC addresses correctly assigned? A table like here would be required: https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses . Note that somebody has already retrieved the assignment from vendor, which I added to the commit message in my version.
* Is there any justification for using 0xee00000 for the total flash size

Please put your answers here and do _not_ update the PR, I don't want to diff everything another time.

**Comment by pgwipeout** [comment]
Good Morning,

Since I'm on vacation to my 'rents where the original device I built for was located, I have successfully tested this on the xr500.
Both the r7800 and the xr500 have successfully been tested.

As requested, I've confirmed that unique mac addresses are being assigned by reading the mac address art tables. They appear to be unique addresses incremented by 1.
For example, this device is:
```
eth0 (wan): fb:c7
eth1 (lan): fb:c6
wlan0 (5g): fb:c8
wlan1 (2.4g): fb:c9
```
This matches the behavior with stock OpenWRT on the r7800.

I can't speak to the original reason why 0xee00000 was used for the original size and I don't have an unmodified stock device to check against.
It goes back to the original pull request, and I'm hesitant to change it since it works.
The only device I have access to is in use and I can't remove it from service.

**Comment by pgwipeout** [comment]
@adschm After looking at your past reviews, I examined the device and yes, I believe the WAN and LAN MACs need to be swapped. Currently LAN is listed as the one on the bottom of the device.
When I get home from vacation I'll check my r7800 as well, it may also need to be swapped.

**Comment by pgwipeout** [comment]
@adschm The r7800 has the correct MAC assigned, so it seems the swapped them in the xr500.

**Comment by nbrtnk** [comment]
Hi,

No idea what I'm doing, but I've got an xr500 with stock firmware.  Here are the MAC address assignments I pulled out:

ethlan HWaddr BC:*:*:*:*:4F
ethwan HWaddr BC:*:*:*:*:50
wifi0 HWaddr BC:*:*:*:*:51
wifi1 HWaddr BC:*:*:*:*:4F 

root@XR500:/# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00c80000 00020000 "qcadata"
mtd1: 00500000 00020000 "APPSBL"
mtd2: 00080000 00020000 "APPSBLENV"
mtd3: 00140000 00020000 "ART"
mtd4: 00140000 00020000 "ART.bak"
mtd5: 00220000 00020000 "kernel"
mtd6: 02fe0000 00020000 "rootfs"
mtd7: 0a000000 00020000 "netgear"
mtd8: 03200000 00020000 "firmware"
mtd9: 00080000 00020000 "crashdump"
mtd10: 00380000 00020000 "language"
mtd11: 00120000 00020000 "config"
mtd12: 00120000 00020000 "pot"
mtd13: 00080000 00020000 "cert"
mtd14: 00010000 00001000 "m25p80"
mtd15: 0005d000 0001f000 "pot.bak"
mtd16: 001b2000 0001f000 "traffic_meter"
mtd17: 001b2000 0001f000 "traffic_meter.bak"
mtd18: 001b2000 0001f000 "dongle"
mtd19: 037b4000 0001f000 "overlay_volume"

output of hexdump -C /dev/mtd3 shows 

00000000  bc * * * * 4f bc *  * * * 50 bc * * *
00000010  * 51

**Comment by pgwipeout** [comment]
Hey that's great @nbrtnk.
Could you do something else, so we can get the absolute addresses of each partition?
The nand partitions are listed at `/sys/firmware/devicetree/base/soc/nand-controller@1ac00000/nand@0/partitions`
If you could run an `ls` in that directory, you should get all the partitions with their addresses.
For example, on OpenWRT the XR500 is:
```
#address-cells     APPSBL@c80000      art@1200000        compatible         name               reserve@e680000
#size-cells        APPSBLENV@1180000  art@1340000        kernel@1480000     qcadata@0          ubi@1880000
```

If you are missing the @ addresses, you can hexdump the `reg` in each partition to get the values, note they are little endian there.

**Comment by nbrtnk** [comment]
/sys/firmware/ is empty.  I tried hexdump and the router hung and now it's bricked.  The usb2 light is on, but that's about it.

**Comment by pgwipeout** [comment]
> /sys/firmware/ is empty. I tried hexdump and the router hung and now it's bricked. The usb2 light is on, but that's about it.

First, what did you hexdump if it was empty?
If the folder didn't populate the syscall would return error file not found.
Have you tried rebooting it?
/sys/firmware is symlinked from /proc/devicetree.
It being empty, combined with the behavior you just described, is exactly why OpenWRT is necessary. Router manufacturers cannot be trusted to make sane software.

**Comment by nbrtnk** [comment]
> First, what did you hexdump if it was empty? If the folder didn't populate the syscall would return error file not found. 
Sorry, I edited my previous reply before posting and reading it again it doesn't make sense.

Basically after I found /sys/firmware was empty I restarted and used dmesg but the message buffer was too small and it just showed the wifi coming up and a bunch of DumaOS stuff.  I then used find to see if I could locate something resembling the nand partition directory, went away and came back to find the router had hung.

Tried the usual restarting, holding down the reset button etc.. but nothing.  Weirdly though it does seem to be responding on some level, looks to be functioning as just a hub/switch.  It's out of warranty, but I guess I could open it up and see if there's a JTAG or some sort of GPIO?

**Comment by Ansuel** [comment]
We bricked another device?

Il giorno gio 7 ott 2021 alle ore 18:38 Adrian Schmutzler <
***@***.***> ha scritto:

**Comment by pgwipeout** [comment]
@nbrtnk Are you located in the US?
If you are unable to make progress on this I'd like to examine the device and will cover conus shipping.

**Comment by pgwipeout** [comment]
@adschm As I've stated previously, the sole device I have access to is in production and unable to return to stock.
And the reason I keep bringing up the r7800 is this is about convention.
There appears to be an established convention for OpenWRT's handling of this, which that device is serving as an example of.

As @nbrtnk was kind enough to post, the MAC table in stock on his device was:
> ethlan HWaddr BC:::::4F
> ethwan HWaddr BC:::::50
> wifi0 HWaddr BC:::::51
> wifi1 HWaddr BC:::::4F

This would make the stock table:
```
use   address   source
LAN   *:4f      art 0x0
WAN   *:50      art 0x6
2g    *:4f      art 0x0
5g    *:51      art 0x12
```

Reusing MACs between devices is a violation of standard, and the OpenWRT script instead looks at art 0x0 and 0x6, then adds 1 and 2 to the value of 0x6 to generate the wifi addresses.
Also it seems that Netgear isn't even following their own convention, as the public MAC which should be the WAN port is attached to the LAN port even on their downstream firmware.

If there is an OpenWRT established convention that the r7800 is following, then we are simply going in circles here.
If you are attempting to establish a convention, then please state as such. Also if so, please tell me you aren't trying to make downstream's illegal implementation that standard.

**Pull request closed without merging**